cpp/include/cudf/detail/utilities/cuda_memcpy.hpp

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf::detail {
+
+enum class copy_kind { PINNED_TO_DEVICE, DEVICE_TO_PINNED, PAGEABLE_TO_DEVICE, DEVICE_TO_PAGEABLE };
```

Comment:
Why is `copy_kind` needed at all? There is exactly one case (pinned, size less than threshold) where you do anything other than pass through to `cudaMemcpyAsync`. You can detect that case with `cudaPtrGetAttributes` and call Thrust for that one case, and just call `cudaMemcpyAsync(cudaMemcpyDefault)` for everything else.

Reply:
It's possible that we'll eventually have a separate threshold for pageable copies, where we copy to a pinned buffer and then `thrust::copy`. @abellina had this in the POC implementation, and IIRC it was helpful even with the extra copy.
I understand current implementation is just a wrapper, I just wanted to leave room for more complex behavior without future changes to the API.

Reply:
OK I see. Does direction affect the choice at all? Could reduce 4 to 2 cases?

Reply:
Reduces to two cases; only the host memory type is specified now.
I can also add an AUTO/DEFAULT option that would call `cudaPointerGetAttributes`. Let me know what you think.